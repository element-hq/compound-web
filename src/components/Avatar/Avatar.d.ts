import React from "react";
type AvatarProps = {
    src?: string;
    id: string;
    name: string;
    type?: "square" | "round";
    className?: string;
    size?: CSSStyleDeclaration["height"];
};
export declare const Avatar: ({ src, id, name, type, className, size, }: AvatarProps) => React.JSX.Element;
export {};
