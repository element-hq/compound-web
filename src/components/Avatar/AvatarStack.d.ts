import React from "react";
type AvatarStackProps = {
    className?: string;
};
/**
 * Renders a stack of avatars and clips the content appropriately.
 *
 * The `type` of avatars should always be set to `round`
 * And all the avatars should have the same size.
 */
export declare const AvatarStack: ({ children, className, }: React.PropsWithChildren<AvatarStackProps>) => React.JSX.Element;
export {};
//# sourceMappingURL=AvatarStack.d.ts.map