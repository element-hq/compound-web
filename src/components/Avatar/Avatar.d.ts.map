{"version":3,"file":"Avatar.d.ts","sourceRoot":"","sources":["../../../../src/components/Avatar/Avatar.tsx"],"names":[],"mappings":"AAiBA,OAAO,KAAmB,MAAM,OAAO,CAAC;AAMxC,KAAK,WAAW,GAAG;IACjB,GAAG,CAAC,EAAE,MAAM,CAAC;IACb,EAAE,EAAE,MAAM,CAAC;IACX,IAAI,EAAE,MAAM,CAAC;IACb,IAAI,CAAC,EAAE,QAAQ,GAAG,OAAO,CAAC;IAC1B,SAAS,CAAC,EAAE,MAAM,CAAC;IACnB,IAAI,CAAC,EAAE,mBAAmB,CAAC,QAAQ,CAAC,CAAC;CACtC,CAAC;AAEF,eAAO,MAAM,MAAM,8CAOhB,WAAW,KAAG,SAAS,QAkCzB,CAAC"}