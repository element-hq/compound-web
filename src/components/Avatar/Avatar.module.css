/*
Copyright 2023  New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.avatar {
  display: inline-block;
  aspect-ratio: 1 / 1;
  width: var(--cpd-avatar-size);
  line-height: var(--cpd-avatar-size); 
  text-align: center;
  font-size: min(calc(var(--cpd-avatar-size) * 0.75), 60px);
  border-radius: var(--cpd-avatar-radius);
  text-transform: uppercase;
  speak: none;
  pointer-events: none;
  font-weight: normal;
  overflow: hidden;
  user-select: none;
}

img.avatar {
  object-fit: cover;
  overflow: hidden; 
}

.avatar:not([src]) {
  /* In the future we'd prefer to pass the HEX code as the data attr
     and use `attr(data-color)` to avoid the style declaration from below
     but this is currently not supported in all browsers */
  background: var(--cpd-avatar-background);
  color: #fff; 
}

.avatar[data-color] {
  --cpd-avatar-background: #368bd6; /* TODO: Replace with a compound color */
}

.avatar[data-color="2"] {
  --cpd-avatar-background: #ac3ba8; /* TODO: Replace with a compound color */
}

.avatar[data-color="3"] {
  --cpd-avatar-background: #03b381; /* TODO: Replace with a compound color */
}

.avatar[data-color="4"] {
  --cpd-avatar-background: #e64f7a; /* TODO: Replace with a compound color */
}

.avatar[data-color="5"] {
  --cpd-avatar-background: #ff812d; /* TODO: Replace with a compound color */
}

.avatar[data-color="6"] {
  --cpd-avatar-background: #2dc2c5; /* TODO: Replace with a compound color */
}

.avatar[data-color="7"] {
  --cpd-avatar-background: #5c56f5; /* TODO: Replace with a compound color */
}

.avatar[data-color="8"] {
  --cpd-avatar-background: #74d12c; /* TODO: Replace with a compound color */
}

.avatar[data-type="round"] {
  --cpd-avatar-radius: 50%;
}

.avatar[data-type="square"] {
  --cpd-avatar-radius: 25%;
}
