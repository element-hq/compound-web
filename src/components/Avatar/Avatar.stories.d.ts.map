{"version":3,"file":"Avatar.stories.d.ts","sourceRoot":"","sources":["../../../../src/components/Avatar/Avatar.stories.tsx"],"names":[],"mappings":";;;;;;;;AAqBA,wBAWkC;AAMlC,eAAO,MAAM,KAAK;;;;;;;EAAoB,CAAC;AAKvC,eAAO,MAAM,MAAM;;;;;;;EAAoB,CAAC;AAKxC,eAAO,MAAM,eAAe;;;;;;;EAAoB,CAAC;AAKjD,eAAO,MAAM,oBAAoB;;;;;;;EAAoB,CAAC"}