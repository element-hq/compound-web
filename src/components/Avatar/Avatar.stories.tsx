/*
Copyright 2023 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from "react";
import { Meta, StoryFn } from "@storybook/react";

import { Avatar as AvatarComponent } from "./Avatar";

export default {
  title: "Avatar",
  component: AvatarComponent,
  tags: ["autodocs"],
  argTypes: {},
  args: {
    name: "Bob",
    size: "64px",
    id: "@bob:example.org",
    src: "/images/__test__/kitten.jpg",
    type: "round",
  },
} as Meta<typeof AvatarComponent>;

const Template: StoryFn<typeof AvatarComponent> = (args) => (
  <AvatarComponent {...args} />
);

export const Round = Template.bind({});
Round.args = {
  type: "round",
};

export const Square = Template.bind({});
Square.args = {
  type: "square",
};

export const Button = Template.bind({});
Button.args = {
  type: "round",
  onClick: () => console.log("clicked!"),
};

export const NoImageFallback = Template.bind({});
NoImageFallback.args = {
  src: "",
};

export const LargeNoImageFallback = Template.bind({});
LargeNoImageFallback.args = {
  src: "",
  size: "128px",
};

const ImageLessCollection: StoryFn<typeof AvatarComponent> = (args) => (
  <>
    <AvatarComponent {...args} id="1" />
    &nbsp;
    <AvatarComponent {...args} id="2" />
    &nbsp;
    <AvatarComponent {...args} id="3" />
    &nbsp;
    <AvatarComponent {...args} id="4" />
    &nbsp;
    <AvatarComponent {...args} id="5" />
    &nbsp;
    <AvatarComponent {...args} id="6" />
    &nbsp;
    <AvatarComponent {...args} id="7" />
    &nbsp;
    <AvatarComponent {...args} id="8" />
    &nbsp;
  </>
);

export const AllAvatars = ImageLessCollection.bind({});
AllAvatars.args = {
  src: "",
  size: "36px",
};
