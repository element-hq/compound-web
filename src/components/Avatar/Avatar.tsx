/*
Copyright 2023  New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import classnames from "classnames";
import React, { Suspense } from "react";
import { getInitialLetter } from "../../utils/string";
import { SuspenseImg } from "../../utils/SuspenseImg";
import styles from "./Avatar.module.css";
import { useUsernameColor } from "./useUsernameColor";

type AvatarProps = {
  src?: string;
  id: string;
  name: string;
  type?: "square" | "round";
  className?: string;
  size?: CSSStyleDeclaration["height"];
};

export const Avatar = ({
  src,
  id,
  name = "",
  type = "round",
  className = "",
  size,
}: AvatarProps): JSX.Element => {
  const color = useUsernameColor(id);
  const style = {
    "--cpd-avatar-size": size,
  } as React.CSSProperties;
  const imagelessAvatar = (
    <span
      role="img"
      aria-label=""
      data-type={type}
      data-color={color}
      className={classnames(styles.avatar, className)}
      style={style}
      title={id}
    >
      {getInitialLetter(name)}
    </span>
  );

  return !src ? (
    imagelessAvatar
  ) : (
    <Suspense fallback={imagelessAvatar}>
      <SuspenseImg
        src={src}
        className={classnames(styles.avatar, className)}
        data-type={type}
        style={style}
        width={size}
        height={size}
        title={id}
      />
    </Suspense>
  );
};
