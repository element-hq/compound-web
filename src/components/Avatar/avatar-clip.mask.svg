<!--
Copyright 2023 New Vector Ltd.

SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
-->

<svg>
  <clipPath id="cpdAvatarClip" clipPathUnits="objectBoundingBox">
    <!-- 
        All values below need to be between 0 and 1
        for objectBoundingBox to work
     -->
    <path d="
        M 0.875 0.16666
        A 0.5 0.5 0 1 0 0.875 0.83333
        Q 0.65 0.5 0.875 0.16666
        Z
     "/>
  </clipPath>
</svg>
