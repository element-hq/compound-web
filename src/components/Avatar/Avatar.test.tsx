/*
Copyright 2023 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { render, waitFor } from "@testing-library/react";
import React from "react";

import { Avatar } from "./Avatar";

const originalImage = global.Image;

describe("Avatar", () => {
  beforeEach(() => {
    global.Image = class extends originalImage {
      constructor() {
        super();
        setTimeout(() => {
          this.onload?.(new Event("load"));
        }, 50);
      }
    };
  });

  afterEach(() => {
    global.Image = originalImage;
  });

  it("renders the image-less avatar", () => {
    const { asFragment } = render(<Avatar name="Bob" id="@bob:example.org" />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("replaces the placeholder with the actual image", async () => {
    const { container } = render(
      <Avatar src="./mock.jpg" name="Bob" id="@bob:example.org" />
    );

    expect(container).toHaveTextContent("B");
    await waitFor(() => expect(document.querySelector("img")).not.toBeNull());
  });

  it("does not split emoji as first letter", () => {
    const { container } = render(
      <Avatar name="ğŸ¤“ John" id="@john:example.org" />
    );
    expect(container).toHaveTextContent("ğŸ¤“");
  });

  it.each([
    ["@bob:example.org", "8"],
    ["@alice:example.org", "3"],
    ["@charlie:example.org", "5"],
    ["@dan:example.org", "8"],
    ["@elena:example.org", "2"],
    ["@fanny:example.org", "1"],
  ])(
    "has a deterministic color based on the id passed to it, %s",
    (id, colorNumber) => {
      const { container } = render(<Avatar name={id} id={id} />);
      expect(container.firstChild).toHaveAttribute("data-color", colorNumber);
    }
  );
});
