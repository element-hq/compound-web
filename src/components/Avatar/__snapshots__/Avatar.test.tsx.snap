// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar renders the image-less avatar 1`] = `
<DocumentFragment>
  <span
    aria-label=""
    class="avatar"
    data-color="8"
    data-type="round"
    role="img"
    title="@bob:example.org"
  >
    B
  </span>
</DocumentFragment>
`;
